src/microsim/schema/emission.py

```diff
@@ -0,0 +1,121 @@
+    # x1000?
+    # this came from calculations elsewhere, and looking at wikipedia
+    # and looking at Nathan Shaner's code
+    # need to double check whether it's still correct with our units
+    ec = ec * 1000
+    return (ec * np.log(10) / AVOGADRO).to("cm^2")  # type: ignore [no-any-return]
+
+
+def energy_per_photon(wavelength: Any) -> pint.Quantity:
+    """Converts a wavelength to energy per photon in J."""
+    wavelength = _ensure_quantity(wavelength, "1nm")
+    return (PLANCK * C / wavelength).to("J")  # type: ignore [no-any-return]
+
+
+def get_excitation_rate(
+    ex_filter_spectrum: Spectrum,
+    fluor: Fluorophore,
+    *,
+    light_power: float = 100,
+) -> Spectrum:
+    """Calculate the number of emission events per fluorophore per second.
+
+    This function takes an excitation spectrum configuration,
+    and the name of a fluorophore, and returns the number of emission events per second
+    per fluorophore.
+    """
+    # get the fluorophore
+    if not (fluor_ex_spectrum := fluor.excitation_spectrum):
+        raise NotImplementedError("Fluorophore has no excitation spectrum.")
+
+    if (ext_coeff := fluor.extinction_coefficient) is None:
+        raise NotImplementedError("Fluorophore has no extinction coefficient.")
+
+    # TODO: derive light power from model
+    irradiance = ex_filter_spectrum * _ensure_quantity(light_power, "W/cm^2")
+    cross_section = fluor_ex_spectrum * ec_to_cross_section(ext_coeff)
+    power_absorbed = cross_section * irradiance
+    excitation_rate = power_absorbed / energy_per_photon(power_absorbed.wavelength)
+    return excitation_rate
+
+    # TODO
+    # if the fluorophore has a lifetime, calculate the effective excitation rate
+    # based on the fraction of time in the excited state
+    # (i.e. at some point, we'll hit ground state depletion and saturation effects)
+    if (lifetime := getattr(fluor, "lifetime", None)) is not None:
+        lifetime = _ensure_quantity(lifetime, "ns")
+        # Calculate the fraction of time in the excited state
+        f_excited = excitation_rate * lifetime
+        # Calculate the fraction of time in the ground state
+        f_ground = 1 - f_excited
+        # Calculate the effective excitation rate
+        excitation_rate = excitation_rate * f_ground
+
+    return excitation_rate
+
+
+def get_emission_events(
+    channel: OpticalConfig,
+    fluor: Fluorophore,
+    *,
+    light_power: float = 100,
+) -> Spectrum:
+    # get the emission events for the given fluorophore
+    if (channel_ex := channel.excitation) is None:
+        raise NotImplementedError("Channel without excitation spectrum?")
+
+    if not (fluor_em_spectrum := fluor.emission_spectrum):
+        raise NotImplementedError("Fluorophore has no emission spectrum.")
+
+    ex_rate = get_excitation_rate(channel_ex.spectrum, fluor, light_power=light_power)
+
+    # convert the excitation rate to an emission rate, by matching the integral
```

Comment:
in another PR, i think it would be good to write out this model in a set of LaTex equations so it's easier to explain how we're modeling excitation and emission